tests/test_frame.py

```diff
@@ -1,640 +1,183 @@
+    mainFrame = p.mainFrame
+    await p.goto(server.cross_process_server / 'empty.html')
+    assert p.mainFrame is mainFrame
+
+@sync
+async def test_attaching(isolated_page, server):
+    p = isolated_page
+    # should detach child frames on navigation
+    attachedFrames = []
+    detachedFrames = []
+    navigatedFrames = []
+    p.on('frameattached', lambda frame: attachedFrames.append(frame))
+    p.on('framedetached', lambda frame: detachedFrames.append(frame))
+    p.on('framenavigated', lambda frame: navigatedFrames.append(frame))
+    await p.goto(server / 'frames/nested-frames.html')
+    assert len(attachedFrames) == 4
+    assert len(detachedFrames) == 0
+    assert len(navigatedFrames) == 5
+
+    # should detach child frames on navigation
+    attachedFrames = []
+    detachedFrames = []
+    navigatedFrames = []
+    await p.goto(server.empty_page)
+    assert len(attachedFrames) == 0
+    # TODO here detachedframes has more than 4 because it's filled with dupes
+    assert len(detachedFrames) == 4
+    assert len(navigatedFrames) == 1
+
+@sync
+async def test_report_frame(isolated_page, server):
+    p = isolated_page
+    # should report frame from-inside shadow DOM
+    await p.goto(server / 'shadow.html')
+    await p.evaluate(
+        """
+        async url => {
+            const frame = document.createElement('iframe');
+            frame.src = url;
+            document.body.shadowRoot.appendChild(frame);
+            await new Promise(x => frame.onload = x);
+        }
+        """,
+        server.empty_page
+    )
+    assert len(p.frames) == 2
+    assert p.frames[1].url == server.empty_page
+
+@sync
+async def test_report_frame_name(isolated_page, server):
+    p = isolated_page
+    await attachFrame(p, server.empty_page, 'theFrameId')
+    await p.evaluate(
+        """
+        url => {
+            const frame = document.createElement('iframe');
+            frame.name = 'theFrameName';
+            frame.src = url;
+            document.body.appendChild(frame);
+            return new Promise(x => frame.onload = x);
+        }
+        """,
+        server.empty_page
+    )
+    assert p.frames[0].name == ''
+    assert p.frames[1].name == 'theFrameId'
+    assert p.frames[2].name == 'theFrameName'
+
+@sync
+async def test_report_frame_parents(isolated_page, server):
+    p = isolated_page
+    await attachFrame(p, server.empty_page, 'frame1')
+    await attachFrame(p, server.empty_page, 'frame2')
+    assert p.frames[0].parentFrame is None
+    assert p.frames[1].parentFrame is p.mainFrame
+    assert p.frames[2].parentFrame is p.mainFrame
+
+@sync
+async def test_frame_reattach(isolated_page, server):
+    p = isolated_page
+    frame1 = await attachFrame(p, server.empty_page, 'frame1')
+    await p.evaluate(
+        """
+        () => {
+            window.frame = document.querySelector('#frame1');
+            window.frame.remove();
+        }
+        """
+    )
+    assert frame1.isDetached is True
```

Comment:
Nitpick, but more idiomatically expressed this could be `assert frame1.isDetached`